### PR TITLE
Regularly purge expired messages

### PR DESCRIPTION
**Pull request by tamf**

### CODE REVIEW

**Review by swongg — COMMENTED** [review]
From the firebase console, I  can see that _deleteExpiredMessages is a function endpoint. I was just curious when this method gets called? Is there a way that this method gets called automatically after certain time intervals? I did a bit of googling but wasn't able to find this.

File: functions/index.js [hunk 1]
```diff
@@ -76,23 +83,54 @@ exports.fetch = functions.https.onRequest(async (req, res) => {
...
+
+  // Delete current batch of documents
+  const batch = admin.firestore().batch();
+  snapshot.docs.forEach((doc) => {
+    batch.delete(doc.ref);
+  });
+  await batch.commit();
+
+  // Repeat
+  process.nextTick(() => {
```
Comment (swongg) [inline functions/index.js:133]: For my learning - what's the purpose of making this method recursive (and deleting in batches of 100 at a time)

File: functions/index.js [hunk 2]
```diff
@@ -76,23 +83,54 @@ exports.fetch = functions.https.onRequest(async (req, res) => {
...
+    deleteQueryBatch(query);
+  });
```
Comment (swongg) [inline functions/index.js:135]: Curious as to what line 133 and line 134 is doing. From my understanding of the deleteQueryBatch method, we pass a reference to a collection of messages where the # of access is <= 0 to the _deleteQueryBatch_ method.
Then we proceed to delete all the message objects from the batch.

batch.commit() is then called to commit all pending operations to the database.

Not sure why we need to call deleteQueryBatch recursively on line 134.

Thanks!

**Review by swongg — APPROVED** [review]

**Comment by swongg** [comment]
Looks good but I'm just curious about some stuff. Comments are located above. Thanks!

**Comment by tamf** [comment]
> From the firebase console, I can see that _deleteExpiredMessages is a function endpoint. I was just curious when this method gets called? Is there a way that this method gets called automatically after certain time intervals? I did a bit of googling but wasn't able to find this.

`_deleteExpiredMessages` is not an exported function so it is not called outside of this file. You can search this file for references of `_deleteExpiredMessages` to see where it is called.

The newly added exported function, `deleteExpiredMessages` is the one you probably meant to ask about. It is called by Google Cloud Scheduler which I have given a cron schedule and this endpoint to hit. This endpoint is slightly different from the fetch/create endpoints in that it is not publicly accessible.

File: functions/index.js (hunk 2 above)
Comment (tamf) [reply functions/index.js:135]: The base case of the recursive function is `batchSize === 0`

There can be multiple batches of messages to delete. The batch size has a max limit which I have defined as a new constant in this file.

**Pull request merged**